staging/src/k8s.io/api/discovery/v1alpha1/types.go

```diff
@@ -57,6 +57,8 @@ type AddressType string
 const (
 	// AddressTypeIP represents an IP Address.
 	AddressTypeIP = AddressType("IP")
+	// AddressTypeFQDN represents a FQDN.
+	AddressTypeFQDN = AddressType("FQDN")
```

Comment:
If people use this and roll back to a prior version, the objects will not pass validation any longer.

Reply:
@lavalamp Good catch, what's the best way to add a type like this? Is it better to add a change like this along with a new API version lie betav1?

Reply:
Or in this case, v1alpha2. But things still have to be round-trippable, so that doesn't really help. The actual way to make this safe is to release a version that understands the new address type, but doesn't permit new instances of it, followed by a next version that both understands and permits usages.

We generally permit alpha APIs to break things, so it might be OK to not do that dance. (@liggitt can confirm?)

Reply:
> If people use this and roll back to a prior version

The primary concern with alpha-level APIs (and fields) is around schema type compatibility, not API server validation. If this was changing the **type** of the field, it should do so in a new version to ensure no client written against the v1alpha1 API got a type error trying to decode an API server response. That API server with that new version should either round-trip it v1alpha1 or drop v1alpha1 entirely (which is [allowed on a release boundary](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-parts-of-the-api)) and release note that existing persisted v1alpha1 endpointslices must be deleted before upgrading.

If the question is just about adding an allowed enum value to validation, that's worth a release note, but not a version bump in my opinion. Full support for upgrade/rollback of alpha/experimental features is not claimed.

> The actual way to make this safe is to release a version that understands the new address type, but doesn't permit new instances of it, followed by a next version that both understands and permits usages.

That is the dance we would require for a field/resource that was beta level or above, correct.

Reply:
> If the question is just about adding an allowed enum value to validation

It's worse than that, the different enum effectively changes the behavior of a string field; the old system only understands IP addresses, now the new one will also permit FQDNs to be stored there.

Reply:
> It's worse than that, the different enum effectively changes the behavior of a string field; the old system only understands IP addresses, now the new one will also permit FQDNs to be stored there.

If it's still type-compatible (and especially if any old data is still valid), that seems like an in-bounds change for an alpha-level API.

Reply:
Yeah, my original note was mostly meant to be informational.

Reply:
Thanks for the clarification @liggitt and @lavalamp! The plan is for this API to graduate to beta in 1.17 but we'd still like to leave the option open of adding additional address types in the future. Would the best approach here be to not prevent the creation of EndpointSlices with unrecognized address types here and maybe add some kind of soft error like an event? Consumers of this API (primarily kube-proxy at this point) would then just have to filter out address types they could not handle (already necessary with this FQDN update).

Reply:
No, that lets people put in land mines (e.g., it's unrecognized today, but tomorrow it's recognized and also invalid because it was entered incorrectly).

Adding new address types in a beta or GA API will require the laborious dance I described.

I recommend that you figure out the types you will eventually want, and put recognition and validation code in place, and just prevent folks from making new instances of the ones you're not ready for. Then you'll have pre-done the first step of the dance and you can enable them in one release.

However! Let me mention another potential failure mode: using these (e.g., via controller) in the same release that you enable them in apiserver. This means that the new controller, should it happen to talk to an old apiserver, will break. In theory, folks should upgrade apiserver first, but in practice, I'm not sure everyone does the upgrade in the right order (in fact, I am sure that some do not do so).

Reply:
It is hard to design for hypothetical as we do not have concrete use cases for now. We entertain a few ideas, but there is no easy way to add more address types after beta. (assuming we can make breaking change in alpha) For the consumer/producer of the EndpointSlice we control, we can safe guard its internals by filtering specific types so that they are ignoring future types. 

It looks inevitable to go thru the 2-release process for  introducing new types in the future.